components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/PublicClientAuthenticator.java

```diff
@@ -0,0 +1,183 @@
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is dedicated for authenticating 'Public Clients'. Public clients do not need a client secret to be
+ * authenticated. This type of authentication is regularly utilised by native OAuth2 clients.
+ */
+public class PublicClientAuthenticator extends AbstractOAuthClientAuthenticator {
+
+    private static Log log = LogFactory.getLog(PublicClientAuthenticator.class);
+
+    /**
+     * Returns the execution order of this authenticator.
+     *
+     * @return Execution place within the order.
+     */
+    @Override
+    public int getPriority() {
+
+        return 200;
+    }
+
+    /**
+     * Authenticates the client.
+     *
+     * @param request                 HttpServletRequest which is the incoming request.
+     * @param bodyParams              Body parameter map of the request.
+     * @param oAuthClientAuthnContext OAuth client authentication context.
+     * @return Whether the authentication is successful or not.
+     * @throws OAuthClientAuthnException
+     */
+    @Override
+    public boolean authenticateClient(HttpServletRequest request, Map<String, List> bodyParams, OAuthClientAuthnContext
+            oAuthClientAuthnContext) {
+
+        return true;
+    }
+
+    /**
+     * Returns whether the incoming request can be authenticated or not using the given inputs.
+     *
+     * @param request    HttpServletRequest which is the incoming request.
+     * @param bodyParams Body parameters present in the request.
+     * @param context    OAuth2 client authentication context.
+     * @return True if can be authenticated, False otherwise.
+     */
+    @Override
+    public boolean canAuthenticate(HttpServletRequest request, Map<String, List> bodyParams, OAuthClientAuthnContext
+            context) {
+
+        String clientId = getClientId(request, bodyParams, context);
+
+        try {
+            if (isClientIdExistsAsParams(bodyParams)) {
+                if (canBypassClientCredentials(context.getClientId())) {
+                    if (clientId != null) {
+                        context.setClientId(clientId);
+                    }
+                    return true;
+                } else {
+                    log.error("The client under the given client ID: " + clientId + " can handle requests" +
+                            " without an authorization header only if the credentials bypass option has been enabled." +
+                            " You can do this by checking the \"Allow authentication without the client secret\"" +
+                            " option in the OAuth application settings page.");
+                }
+            } else {
+                log.error("Application with the given client ID: " + clientId + "is not found");
+            }
+        } catch (InvalidOAuthClientException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Client ID could not be retrieved from the request.", e);
```

Comment:
Error : "Error in retrieving an Application (Service Provider) with client ID : X", e
as this is a exception, needs the trace.